### PR TITLE
60 index selector non negative

### PR DESCRIPTION
**Pull request by zwerdlds**
## Issue
This PR implements the non-negative index selector.

Resolves: #60 

## Checklist
- [x] I ran `just verify` locally and it succeeded.
- [x] Issue was given <span style="color: #FF4400">go ahead</span> and is linked above
- [x] Unit tests for my changes are included

### CODE REVIEW

File: crates/rsonpath-lib/src/query/parser.rs [hunk 1]
```diff
@@ -374,6 +421,61 @@ mod tests {
...
+        super::array_index_selector()(input).unwrap_err();
+    }
+
+    #[test]
+    fn two_pow_fiftythree_minus_one_array_index() {
+        let input = "[9007199254740991]";
+
+        let result = super::array_index_selector()(input);
+
+        assert_eq!(result, Ok(("", 9007199254740991.try_into().unwrap())));
```
Comment (V0ldek) [inline crates/rsonpath-lib/src/query/parser.rs:469]: This is giving me a warning from clippy about long literals:

```suggestion
        assert_eq!(result, Ok(("", 9_007_199_254_740_991.try_into().unwrap())));
```

**Review by V0ldek — CHANGES_REQUESTED** [review]
The error message for too large numbers is less than ideal:

image

Let's leave it as is, but I'll create an issue for improving the messages.

File: crates/rsonpath-lib/src/query.rs [hunk 2]
```diff
@@ -463,4 +517,9 @@ mod tests {
 
         assert_eq!(h1, h2);
     }
+
+    #[test]
+    fn index_ulimit_sanity_check() {
+        assert_eq!(9007199254740991, ARRAY_INDEX_ULIMIT);
+    }
```
[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query.rs:524]: Agreed;  This seems to stem from my misunderstanding regarding the inclusivity described previously

[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query.rs:524]: I added some tests in query to validate the conversion at the bounds.

File: crates/rsonpath-lib/src/query/parser.rs [hunk 3]
```diff
@@ -224,6 +232,45 @@ fn label_character<'a>() -> impl Parser<'a, char> {
...
+fn array_index_child_selector<'a>() -> impl Parser<'a, Token<'a>> {
+    map(
+        alt((array_index_selector(), dot_index_selector())),
+        Token::ArrayIndex,
+    )
+}
+
+fn dot_index_selector<'a>() -> impl Parser<'a, NonNegativeArrayIndex> {
+    preceded(char('.'), array_index_selector())
+}
```
[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query/parser.rs:244]: I've removed the incorrect selector.

File: crates/rsonpath-lib/src/query/parser.rs [hunk 4]
```diff
@@ -224,6 +232,45 @@ fn label_character<'a>() -> impl Parser<'a, char> {
...
+fn parsed_array_index<'a>() -> impl Parser<'a, u64> {
+    map_res(length_limited_array_index(), str::parse)
+}
+
+fn length_limited_array_index<'a>() -> impl Parser<'a, &'a str> {
+    map_res(digit1, |cs: &str| {
+        // When unwrap goes const, this calc can be moved to a const; there might be a feature flag we can enable too
+        ARRAY_INDEX_ULIMIT
+            .checked_ilog10()
+            .ok_or(ArrayIndexError::UpperLimitLengthCalculationError)
```
[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query/parser.rs:263]: I guess you're right that its pretty unlikely the bound will change, which is really the only reason this is here.  Plus if it does it should get caught in at least one test.  Changed per se.

**Review by V0ldek — APPROVED** [review]

File: crates/rsonpath-lib/src/query/parser.rs [hunk 5]
```diff
@@ -255,19 +248,14 @@ fn parsed_array_index<'a>() -> impl Parser<'a, u64> {
...
-            .checked_ilog10()
-            .ok_or(ArrayIndexError::UpperLimitLengthCalculationError)
-            .and_then(|max_index_b10_len| {
-                if cs.len() > ((max_index_b10_len as usize) + 1) {
-                    Err(ArrayIndexError::ExceedsUpperLimitError(cs.to_owned()))
-                } else {
-                    Ok(cs)
-                }
-            })
+        if cs.len() > (ARRAY_INDEX_ULIMIT_B10_DIGIT_CT + 1) {
```
[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query/parser.rs]: I renamed it to ARRAY_INDEX_ULIMIT_BASE_10_DIGIT_COUNT

**Comment by zwerdlds** [comment]
There was a minor change to an unexpected part of the system in f62de26c0f511a65ccbdd23f1514d7aff4c43ac6

**Review by V0ldek — APPROVED** [review]

**Pull request merged**